xkafka/options.go

```diff
@@ -0,0 +1,82 @@
+package xkafka
+
+import (
+	"time"
+
+	"github.com/confluentinc/confluent-kafka-go/kafka"
+)
+
+type consumerOptions struct {
+	topics       []string
+	brokers      []string
+	configMap    kafka.ConfigMap
+	errorHandler ErrorHandler
+
+	metadataRequestTimeout time.Duration
+	pollTimeout            time.Duration
+	shutdownTimeout        time.Duration
+}
+
+func defaultConsumerOptions() consumerOptions {
+	return consumerOptions{
+		topics:                 []string{},
+		brokers:                []string{},
+		configMap:              kafka.ConfigMap{},
+		errorHandler:           NoopErrorHandler,
+		metadataRequestTimeout: 10 * time.Second,
+		pollTimeout:            10 * time.Second,
+		shutdownTimeout:        1 * time.Second,
+	}
+}
+
+// ConsumerOption enables consumer configuration.
+type ConsumerOption func(*consumerOptions)
```

Comment:
nit: While I think we can make this change later as well, instead of making this a named `func(*consumerOptions)` we can go for a [package-only implementable interface](https://github.com/gojek/courier-go/blob/5f1e90a9bea274dc818b8347fe9f4b1d75c341c6/client_options.go#L12-L20) right away.

This will allow to specify typed values. Example:

```go
type ConsumerOption interface{ apply(*consumerOptions) }

type Topics []string

func (ts Topics) apply(o *consumerOptions) { o.topics = ts }

type optionFunc func(*consumerOptions)

func (f optionFunc) apply(o *consumerOptions) { f(o) }

func WithKafkaConfig(key string, value interface{}) ConsumerOption {
	return optionFunc(func(o *consumerOptions) {
		_ = o.configMap.SetKey(key, value)
	})
}
```

Reply:
Got it. Just curious- what advantage does this provide over named func?

Reply:
Advantage here is, type information is retained.
And types themselves can be used as `Option(s)`

Example:
```go
// WithErrorHandler ConsumerOption is no longer required as we can add apply method directly to ErrorHandler.
func WithErrorHandler(fn ErrorHandler) ConsumerOption {
	return func(o *consumerOptions) {
		o.errorHandler = fn
	}
}

type ErrorHandler func(err error) error

// ErrorHandler itself is a ConsumerOption now.
func (h ErrorHandler) apply(o *consumerOptions) { o.errorHandler = h }
```

Also, when dealing with wrapping use-cases, it gives you the ability to modify the options because you now have the type information retained.
Example:

```go
func updateErrorHandler(opts []xkafka.ConsumerOption) []xkafka.ConsumerOption {
	var res []xkafka.ConsumerOption

	for _, opt := range opts {
		switch opt := opt.(type) {
		case xkafka.ErrorHandler:
			res = append(res, xkafka.ErrorHandler(func(err error) error {
				baseErr := opt(err);
				if errors.Is(baseErr, SomeErrType) {
					return nil
				}
				return baseErr;
			}))
		default:
			res = append(res, opt)
		}
	}

	return res
}
```